js_modules/dagster-ui/packages/ui-core/src/assets/RecentUpdatesTimeline.tsx

```diff
@@ -0,0 +1,209 @@
+  Box,
+  Caption,
+  Colors,
+  Icon,
+  Popover,
+  Subtitle2,
+  Tag,
+  useViewport,
+} from '@dagster-io/ui-components';
+import {useMemo} from 'react';
+import {Link} from 'react-router-dom';
+import styled from 'styled-components';
+
+import {RunlessEventTag} from './RunlessEventTag';
+import {assetDetailsPathForKey} from './assetDetailsPathForKey';
+import {isRunlessEvent} from './isRunlessEvent';
+import {AssetKey} from './types';
+import {useRecentAssetEvents} from './useRecentAssetEvents';
+import {Timestamp} from '../app/time/Timestamp';
+import {AssetRunLink} from '../asset-graph/AssetRunLinking';
+import {RunStatusWithStats} from '../runs/RunStatusDots';
+import {titleForRun} from '../runs/RunUtils';
+import {useFormatDateTime} from '../ui/useFormatDateTime';
+
+const MIN_TICK_WIDTH = 5;
+
+export const RecentUpdatesTimeline = ({
+  assetKey,
+  materializations,
+}: {
+  assetKey: AssetKey;
+  materializations: ReturnType<typeof useRecentAssetEvents>['materializations'];
+}) => {
+  const {containerProps, viewport} = useViewport();
+  const widthAvailablePerTick = viewport.width / 100;
+
+  const tickWidth = Math.max(widthAvailablePerTick, MIN_TICK_WIDTH);
+
+  const buckets = Math.floor(viewport.width / tickWidth);
+
+  const endTimestamp = parseInt(materializations[0]!.timestamp);
+  const startTimestamp = parseInt(materializations[materializations.length - 1]!.timestamp);
+  const range = endTimestamp - startTimestamp;
+
+  const bucketedMaterializations = useMemo(() => {
+    if (!viewport.width) {
+      return [];
+    }
+    const firstPassBucketsArray: Array<{
+      number: number;
+      materializations: typeof materializations;
+    }> = new Array(buckets);
+
+    materializations.forEach((materialization) => {
+      const bucketNumber = Math.floor(
+        ((parseInt(materialization.timestamp) - startTimestamp) / range) * buckets,
+      );
+      firstPassBucketsArray[bucketNumber] = firstPassBucketsArray[bucketNumber] || {
+        number: bucketNumber,
+        materializations: [] as typeof materializations,
+      };
+      firstPassBucketsArray[bucketNumber]!.materializations.push(materialization);
+    });
+
+    const secondPassBucketsArray: Array<{
+      start: number;
+      end: number;
+      materializations: typeof materializations;
+    }> = [];
+
+    firstPassBucketsArray.forEach((bucket) => {
+      const lastBucket = secondPassBucketsArray[secondPassBucketsArray.length - 1];
+      if (!lastBucket || lastBucket.end !== bucket.number - 1) {
+        secondPassBucketsArray.push({
+          start: bucket.number,
+          end: bucket.number,
+          materializations: bucket.materializations,
+        });
+      } else {
+        lastBucket.end = bucket.number;
+        lastBucket.materializations = [...lastBucket.materializations, ...bucket.materializations];
+      }
+    });
+    return secondPassBucketsArray;
+  }, [viewport.width, buckets, materializations, startTimestamp, range]);
```

Comment:
Is it possible for this to reuse the bucketing and pixel overlap detection from `batchRunsForTimeline.tsx`? No problem if the file and functions need to be renamed, but I think it should be generic and not specific to runs, as long as there are starts and ends to each "run" object.

Reply:
@hellendag There are no real start/end times. When I bucket each event I give it the start/timestamp of that bucket. I actually keep the concept of buckets around everywhere since it simplifies things a little.